flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalogBase.java

```diff
@@ -318,13 +320,44 @@ public void alterTable(ObjectPath tablePath, CatalogBaseTable newCatalogTable, b
 					throw new TableNotExistException(catalogName, tablePath);
 				}
 			} else {
-				// TODO: [FLINK-12452] alterTable() in all catalogs should ensure existing base table and the new one are of the same type
+				Table oldTable = getHiveTable(tablePath);
+				TableType oldTableType = TableType.valueOf(oldTable.getTableType());
+
+				// Flink cannot alter existing Hive tables in types that Flink cannot create for now,
+				// including EXTERNAL_TABLE, INDEX_TABLE and MATERIALZED_VIEW.
+				// Note that though Flink cannot create them, it's able to read most parts of their metadata,
+				// like schema and properties, thru existing mechanism, but not their unique metadata.
+				// Also note that add capability of creating EXTERNAL_TABLE should be fairly easy to do, e.g. with a table property
+				// "table_type=external", we just haven't done it yet.
+				// INDEX_TABLE and MATERIALZED_VIEW need more research.
+
+				if (oldTableType == TableType.EXTERNAL_TABLE
```

Comment:
Instead of listing types that we don't support, it's probably better to check if the type is what we do support.

Also event though Flink is not able to create external table yet, that's probably due to the limitation of the DLL and I don't see why we should disallow user to modify it.

It seems okay if we say we don't support other types yet.

Reply:
Actually I listed both here just to make it more clear on what we do and do not support right now.

Yeah, as mentioned in the comment, external table is not supported only because we haven't done it yet, not because DDL limitation. Enabling external table will be on its own PR.

Reply:
1. Saying in the comment is fine. I meant when we check in the code, it's better to check against something we support rather than checking against something we don't . The difference is that there might be additional things coming in the future that we don't support. We can avoid code changes if we don't plan to support it.

2. I understand external table cannot be created in Flink. The reason is the limitation of DDL. However, altering an external table should already work. I don't see why we need to exclude it now. Let me know if my judgement is inaccurate.

Reply:
1. I think the current code is better since it checks what we've supported, what we plan to support, what we don't know (the "additional things coming in the future" you mentioned)
2. we don't support it yet. external table has additional properties we need to parse and check